### PR TITLE
[warning] Add warning when values are not being reduced

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

After discussions with @awaelchli @justusschock @SeanNaren, we decided to add a warning on `epoch_end` when a given metric isn't reduced properly across processes.
The warning gives 2 information:
- use `self.log(sync_dist=True)` to reduce the values with mean
- Implement a Metric if the mean reduction is not appropriate

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]
I am a bit out of context but looks fine to me

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 1]
```diff
@@ -104,8 +121,26 @@ def get_batch_log_metrics(self, *args, **kwargs):
...
+        ):
+            for non_metric_key in opt_metric.get_non_metrics_keys():
+                if non_metric_key in metrics_to_log and non_metric_key not in warning_cache.warned_metrics:
+                    metric = self._all_gather_fn(metrics_to_log[non_metric_key])
+                    if any(metric[0] != m for m in metric[1:]):
+                        warning_cache.warn(
+                            f"The value associated to the key {non_metric_key}: {metric.cpu().tolist()} "
+                            "doesn't appear to be the same accross all processes. "
+                            "HINT: One could either do: `self.log(..., sync_dist=True, sync_fn=torch.mean)`"
+                            " to force mean reduction across processes which can be inaccurate or implement"
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:139]: Why can it be inaccurate?

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 2]
```diff
@@ -104,8 +121,26 @@ def get_batch_log_metrics(self, *args, **kwargs):
     def run_epoch_func(self, results, opt_metric, func_name, *args, **kwargs) -> None:
         if not isinstance(opt_metric, Result):
             raise Exception("The provided opt_metric should be a Result Object. Something is wrong")
+
         func = getattr(opt_metric, func_name)
         metrics_to_log = func(*args, add_dataloader_idx=self.has_several_dataloaders, **kwargs)
+        if (
+            torch.distributed.is_available() and torch.distributed.is_initialized()
+            and self._all_gather_fn.__self__.trainer.world_size > 1
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:129]: is this really the easiest way to check the world size?

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:139]: It depends on the metric the user is logging. Not all logged values can be properly reduced by the mean op. Therefore, we provide a warning if we detect the reduction didn't happen.

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:129]: I tried, but I had some issues with windows using get_world_size utils from torch. Any idea ?

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:129]: Let me try to find a better way.

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 3]
```diff
@@ -104,8 +121,26 @@ def get_batch_log_metrics(self, *args, **kwargs):
...
+            for non_metric_key in opt_metric.get_non_metrics_keys():
+                if non_metric_key in metrics_to_log and non_metric_key not in warning_cache.warned_metrics:
+                    metric = self._all_gather_fn(metrics_to_log[non_metric_key])
+                    if any(metric[0] != m for m in metric[1:]):
+                        warning_cache.warn(
+                            f"The value associated to the key {non_metric_key}: {metric.cpu().tolist()} "
+                            "doesn't appear to be the same accross all processes. "
+                            "HINT: One could either do: `self.log(..., sync_dist=True, sync_fn=torch.mean)`"
+                            " to force mean reduction across processes which can be inaccurate or implement"
+                            " a `pytorch_lightning.metrics.Metric`"
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:140]:
```suggestion
                            " a `torchmetrics.Metric`"
```

**Review by justusschock — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
